### PR TITLE
CNV automation: Classes and helper functions - Virtctl, VirtualMachine, VirtualMachineInstance and VM workloads

### PR DESCRIPTION
**Pull request by PrasadDesala**
OpenShift Virtualization (CNV) is an add-on to OCP that allows running and managing virtual machine workloads alongside container workloads. The ODF can be used in conjunction with CNV to provide persistent storage for virtual machines.

This PR handles creation of necessary classes and helper functions for VM workload management.

- [x] Virtctl  - Virtctl is a commandline utility used for managing and interacting with virtual machines in a Kubernetes cluster. It provides a set of commands to control virtual machines such as starting, stopping, SSH, SCP and obtaining information about a VM.
- [x] VirtualMachine  - VirtualMachine is a Kubernetes CR representing a virtual machine. It defines the desired state and configuration
- [x] VirtualMachineInstance  - VirtualMachine is a Kubernetes CR representing a virtual machine instance. It represents an instance of a virtual machine
- [x]  VM referenced by a standalone PVC
- [x] VM workload -  Centos

### CODE REVIEW

File: ocs_ci/helpers/cnv_helpers.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import logging
+
+from ocs_ci.helpers.helpers import create_unique_resource_name, create_resource
+from ocs_ci.ocs import constants
+from ocs_ci.utility import templating
+
+logger = logging.getLogger(__name__)
+
+
+def create_vm(
```
Comment (avd-sagare) [inline ocs_ci/helpers/cnv_helpers.py:13]: can we add in factory function?

Comment (PrasadDesala) [reply ocs_ci/helpers/cnv_helpers.py:17]: Yes, I have plans for it. will create a factory in the next PR

File: ocs_ci/helpers/cnv_helpers.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        CommandFailed: If an error occurs during the creation of the VM
+
+    """
+    vm_dict_path = vm_dict_path if vm_dict_path else constants.CNV_VM_CENTOS_YAML
+    vm_data = templating.load_yaml(vm_dict_path)
+    vm_name = vm_name if vm_name else create_unique_resource_name("test", "vm")
+    # data volume is created with the same vm name
+    vm_data["metadata"]["name"] = vm_name
+    vm_data["metadata"]["namespace"] = namespace
+    vm_data["spec"]["dataVolumeTemplates"][0]["metadata"]["name"] = vm_name
```
Comment (kgoldbla) [inline ocs_ci/helpers/cnv_helpers.py:47]: I noticed you are using datavolume templates which won't be supported in 4.15.0. For 4.15.0 we need the stand alone PVC which is referenced by the vm

Comment (kgoldbla) [reply ocs_ci/helpers/cnv_helpers.py:47]: An example of the stand alone PVC referenced by the vm can be seen here:
https://github.com/kgoldbla/ocm-kubevirt-samples/blob/vm9-application/vm-standalone-pvc-odr-regional/pvc.yaml

File: ocs_ci/templates/cnv-vm-workload/vm_fedora.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: fedora
+
+spec:
+  dataVolumeTemplates:
```
Comment (kgoldbla) [inline ocs_ci/templates/cnv-vm-workload/vm_fedora.yaml:7]: You are using datavolume templates which won't be supported in 4.15.0. For 4.15.0 we need the stand alone PVC which is referenced by the vm

An example of the stand alone PVC referenced by the vm can be seen here:
https://github.com/kgoldbla/ocm-kubevirt-samples/blob/vm9-application/vm-standalone-pvc-odr-regional/pvc.yaml

File: ocs_ci/templates/cnv-vm-workload/vm_centos.yaml [hunk 4]
```diff
@@ -0,0 +1,76 @@
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: centos-stream8
+  namespace: default
+
+spec:
+  dataVolumeTemplates:
```
Comment (kgoldbla) [inline ocs_ci/templates/cnv-vm-workload/vm_centos.yaml:8]: I noticed you are using datavolume templates which won't be supported in 4.15.0. For 4.15.0 we need the stand alone PVC which is referenced by the vm

An example of the stand alone PVC referenced by the vm can be seen here:
https://github.com/kgoldbla/ocm-kubevirt-samples/blob/vm9-application/vm-standalone-pvc-odr-regional/pvc.yaml

**Review by kgoldbla — CHANGES_REQUESTED** [review]

File: ocs_ci/templates/cnv-vm-workload/vm_centos.yaml (hunk 4 above)
Comment (PrasadDesala) [reply ocs_ci/templates/cnv-vm-workload/vm_centos.yaml:8]: Done

File: ocs_ci/templates/cnv-vm-workload/vm_fedora.yaml (hunk 3 above)
Comment (PrasadDesala) [reply ocs_ci/templates/cnv-vm-workload/vm_fedora.yaml]: Done

File: ocs_ci/helpers/cnv_helpers.py (hunk 2 above)
Comment (PrasadDesala) [reply ocs_ci/helpers/cnv_helpers.py]: Done

**Review by parikshithb — APPROVED** [review]
lgtm

File: tests/functional/workloads/cnv/test_vm_lifecycle_and_io.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        1) Create a VM using a standalone PVC
+            a) Create a cdi source with a registry url pointing to the source image
+            b) Create a PVC using this source image backed with an odf storageclass
+            c) Create a secret using a statically manged public SSH key and add this secret name to the VM spec for ssh
+            d) Create a VM using the above PVC
+        2) Start the VM using virtctl command and wait for the VM to reach running state
+        3) SSH to the VM and create some data on the PVC mount point
+        4) SCP that create data in step-3 to localmachine
+        5) Stop the VM
+        6) Delete the VM (as part of teardown)
```
Comment (kgoldbla) [inline tests/functional/workloads/cnv/test_vm_lifecycle_and_io.py:57]: Same here, the vm deleted here will be recreated when reconciled in the github repository

File: tests/functional/workloads/cnv/test_vm_single_disk_hot_plugging_unplugging.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            command="dd if=/dev/zero of=/dd_file.txt bs=1024 count=102400"
+        )
+        pvc_obj = create_pvc(
+            sc_name=constants.DEFAULT_CNV_CEPH_RBD_SC,
+            namespace=self.proj_obj.namespace,
+            size=20,
+            access_mode=constants.ACCESS_MODE_RWX,
+            volume_mode=constants.VOLUME_MODE_BLOCK,
+        )
+        self.vm_obj.addvolme(volume_name=pvc_obj.name)
```
Comment (kgoldbla) [inline tests/functional/workloads/cnv/test_vm_single_disk_hot_plugging_unplugging.py:62]: I believe hot plugging disks to the vm is also something that should be handled via the github repository.

**Review by kgoldbla — CHANGES_REQUESTED** [review]

File: ocs_ci/helpers/cnv_helpers.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    wait_for_vm_boot=True,
+    vm_dict_path=None,
+):
+    """
+    Create a Virtual Machine (VM) in the specified namespace using a standalone Persistent Volume Claim (PVC)
+
+    Args:
+        namespace (str): The namespace in which to create the VM.
+        vm_name (str): Name for the VM. If not provided, a unique name will be generated.
+        existing_data_volume_name (str): Existing DataVolume name to attach to the VM
```
[the remark this answers is not in the document]
Comment (PrasadDesala) [reply ocs_ci/helpers/cnv_helpers.py:37]: I have modified the func to create a VM using , standalone PVC. Looks like I have missed correcting the doc string.

File: ocs_ci/templates/cnv-vm-workload/vm-standalone-pvc/vm.yaml [hunk 8]
```diff
@@ -0,0 +1,61 @@
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: centos-stream8
+  namespace: default
+spec:
+  running: true
```
[the remark this answers is not in the document]
Comment (PrasadDesala) [reply ocs_ci/templates/cnv-vm-workload/vm-standalone-pvc/vm.yaml:7]: These are the core CNV (not related to DR alone) classes and helper functions for VM workload management. The operations that are specific to MDR will be implemented in a separate PR.

File: tests/functional/workloads/cnv/test_vm_lifecycle_and_io.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        """
+        This test performs the VM lifecycle operations and IO
+
+        Steps:
+        1) Create a VM using a standalone PVC
+            a) Create a cdi source with a registry url pointing to the source image
+            b) Create a PVC using this source image backed with an odf storageclass
+            c) Create a secret using a statically manged public SSH key and add this secret name to the VM spec for ssh
+            d) Create a VM using the above PVC
+        2) Start the VM using virtctl command and wait for the VM to reach running state
```
[the remark this answers is not in the document]
Comment (PrasadDesala) [reply tests/functional/workloads/cnv/test_vm_lifecycle_and_io.py:53]: same as above. These are the core CNV (not related to DR alone) classes and helper functions for VM workload management. The operations that are specific to MDR will be implemented in a separate PR.

File: tests/functional/workloads/cnv/test_vm_lifecycle_and_io.py (hunk 5 above)
Comment (PrasadDesala) [reply tests/functional/workloads/cnv/test_vm_lifecycle_and_io.py:57]: This is a basic CNV test not specific to DR

File: tests/functional/workloads/cnv/test_vm_single_disk_hot_plugging_unplugging.py (hunk 6 above)
Comment (PrasadDesala) [reply tests/functional/workloads/cnv/test_vm_single_disk_hot_plugging_unplugging.py:62]: This is a basic CNV test not specific to DR

File: ocs_ci/helpers/cnv_helpers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (PrasadDesala) [reply ocs_ci/helpers/cnv_helpers.py]: fixed doc string

File: ocs_ci/templates/cnv-vm-workload/vm-standalone-pvc/vm.yaml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+      terminationGracePeriodSeconds: 180
+      volumes:
+        -  name: rootdisk
+           persistentVolumeClaim:
+             claimName: sample-vm-pvc
+        - cloudInitNoCloud:
+            userData: |-
+              #cloud-config
+              user: centos
+              password: 7lua-f8o8-nk3t
```
Comment (shylesh) [inline ocs_ci/templates/cnv-vm-workload/vm-standalone-pvc/vm.yaml:59]: Exposing anytype of plaintext username/passwords is not a good practice.

File: ocs_ci/ocs/cnv/virtual_machine_instance.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+from ocs_ci.ocs.resources.pod import Pod
+from ocs_ci.ocs.resources.pod import get_pods_having_label
+from ocs_ci.utility.utils import TimeoutSampler
+from ocs_ci.ocs import constants
+
+
+logger = logging.getLogger(__name__)
+
+
+class VirtualMachineInstance:
```
Comment (shylesh) [inline ocs_ci/ocs/cnv/virtual_machine_instance.py:17]: @PrasadDesala I am not able to clearly understand the difference between VirtualMachine class and VirtualMachineInstance class. To me it looks mostly similar (maybe I lack complete understanding of how CNV + ODF works), operations looks similar and VirtualMachine class functions appears as if they are operating on actual instances. Could you please clarify.

Comment (PrasadDesala) [reply ocs_ci/ocs/cnv/virtual_machine_instance.py:17]: You are right! VM defines the desired state and configuration and it is not associated with a specific pod or process, whereas VMI represents an instance of a running virtual machine.
A few  of the operation calls like delete, start, pause look to be same but it's behavior and the internal calls will be different from each other.
For instance,if we want to stop a VM,  the VM will go to stopped state only after the VMI deletes its associated virt_launcher pod and vmi resource. We need to have separate functions as each maintains it's own crds.

File: ocs_ci/templates/cnv-vm-workload/vm-standalone-pvc/vm.yaml (hunk 10 above)
Comment (PrasadDesala) [reply ocs_ci/templates/cnv-vm-workload/vm-standalone-pvc/vm.yaml:59]: Only the user who secret is added to the VM during the VM creation and has perms to the openshift console can access this VM. But, Yes I agree with you that this way is not a good practice. This is one of the way to send the userdata using the cloud-config, probably I will try to explore the other options and see if we can mask these details.

File: ocs_ci/ocs/cnv/virtctl.py [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+            command = f"sudo {command}"
+
+        mandatory_flags = [
+            f"--username={username}",
+            "--port=22",  # Default port for VM
+            f'-c "{command}"',
+        ]
+
+        if identity_file:
+
```
[the remark this answers is not in the document]
Comment (PrasadDesala) [reply ocs_ci/ocs/cnv/virtctl.py]: Done

**Review by shylesh — APPROVED** [review]
LGTM

**Review by parikshithb — APPROVED** [review]

**Pull request merged**